### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by konopka90**
Hi, I made some refactoring. Most important changes:
 - EngineInputManager is subject and EngineMain, World are listeners
 - EngineInputManager doesn't need EngineMain dependency
 - Some game logic from EngineInputManager moved to World

### CODE REVIEW

**Comment by wheybags** [comment]
Hey, sorry I haven't had a chance to review this yet, I will get to it soon!

File: components/input/hotkey.cpp [hunk 1]
```diff
@@ -14,17 +14,17 @@ namespace Input
         alt = false;
     }
 
-    Hotkey::Hotkey(const char *name)
-    {
-        std::string sname = name;
+    Hotkey::Hotkey(const char *name) : Hotkey(std::string(name)) {}
```
Comment (wheybags) [inline components/input/hotkey.cpp]: Unnecessary IIRC, I think c++ will do this conversion itself

**Comment by wheybags** [comment]
needed
```
diff --git a/apps/freeablo/faworld/world.cpp b/apps/freeablo/faworld/world.cpp
index 5fe5f3e..c0c42a9 100644
--- a/apps/freeablo/faworld/world.cpp
+++ b/apps/freeablo/faworld/world.cpp
@@ -8,6 +8,7 @@
 #include "../engine/enginemain.h"
 #include "actorstats.h"
 #include "monster.h"
+#include <iostream>

 namespace FAWorld
 {
diff --git a/components/input/hotkey.cpp b/components/input/hotkey.cpp
index bfd73e6..d0f2e8d 100644
--- a/components/input/hotkey.cpp
+++ b/components/input/hotkey.cpp
@@ -1,4 +1,5 @@
 #include <string>
+#include <cctype>
 #include <settings/settings.h>
 #include <misc/boost_python.h>
 #include "hotkey.h"
``` 
to compile with msvc, other than that and the other comment I made, looks good!

I should really set up a windows vm running jenkins or something...

File: components/input/hotkey.cpp (hunk 1 above)
Comment (konopka90) [inline components/input/hotkey.cpp]: Probably yes but as I remember this ctor is used with Python and I don't if ctor(std::string) would be working properly.

Comment (wheybags) [inline components/input/hotkey.cpp]: Ah, ok, well it's not a big issue anyway

**Comment by mewmew** [comment]
> I should really set up a windows vm running jenkins or something...

You may want to take a look at https://www.appveyor.com/

For an example config (in Go, but should be adjustable to C++), see https://github.com/EngoEngine/engo/blob/master/appveyor.yml

**Comment by wheybags** [comment]
I actually just saw that, will set it up soon, thanks!

**Comment by konopka90** [comment]
> diff --git a/apps/freeablo/faworld/world.cpp b/apps/freeablo/faworld/world.cpp
index 5fe5f3e..c0c42a9 100644
--- a/apps/freeablo/faworld/world.cpp
+++ b/apps/freeablo/faworld/world.cpp
@@ -8,6 +8,7 @@
 #include "../engine/enginemain.h"
 #include "actorstats.h"
 #include "monster.h"
+#include \<iostream>

Please place includes from std library at the top of header file, it is good habit :)

File: apps/freeablo/engine/input.cpp [hunk 2]
```diff
@@ -1,36 +1,52 @@
...
+                std::bind(&EngineInputManager::mouseClick, this, ph::_1, ph::_2, ph::_3),
+                std::bind(&EngineInputManager::mouseRelease, this, ph::_1, ph::_2, ph::_3),
+                std::bind(&EngineInputManager::mouseMove, this, ph::_1, ph::_2),
+                FARender::Renderer::get()->getRocketContext())
+    {
+        mHotkeys[QUIT] = Input::Hotkey(keyboardActionToString(QUIT));
+        mHotkeys[NOCLIP] = Input::Hotkey(keyboardActionToString(NOCLIP));
+        mHotkeys[CHANGE_LEVEL_UP] = Input::Hotkey(keyboardActionToString(CHANGE_LEVEL_UP));
+        mHotkeys[CHANGE_LEVEL_DOWN] = Input::Hotkey(keyboardActionToString(CHANGE_LEVEL_DOWN));
+        mHotkeys[TOGGLE_CONSOLE] = Input::Hotkey(keyboardActionToString(TOGGLE_CONSOLE));
```
Comment (konopka90) [inline apps/freeablo/engine/input.cpp]: Should be changed to array of enums and for-loop

File: apps/freeablo/engine/input.cpp [hunk 3]
```diff
@@ -73,51 +89,59 @@ namespace Engine
...
-                else
-                    nextLevelIndex = world->getCurrentLevel()->getNextLevel();
-
-                world->setLevel(nextLevelIndex);
+                notifyKeyboardObservers(CHANGE_LEVEL_UP);
+            }
+            else if (hotkey == getHotkey(CHANGE_LEVEL_DOWN))
+            {
+                notifyKeyboardObservers(CHANGE_LEVEL_DOWN);
+            }
```
Comment (konopka90) [inline apps/freeablo/engine/input.cpp]: Also should be changed to array of enums and for-loop

Comment (konopka90) [inline apps/freeablo/engine/input.cpp]: or better: just use set of enums and simple check set.find(hotkey) != set.end() { notify(..) }

File: apps/freeablo/engine/threadmanager.cpp [hunk 4]
```diff
@@ -44,11 +44,11 @@ namespace Engine
             auto now = std::chrono::system_clock::now();
             numFrames++;
             
-            size_t dur = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch() - last.time_since_epoch()).count();
+            size_t duration = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch() - last.time_since_epoch()).count();
             
-            if(dur >= 1000)
+            if(duration >= 1000)
```
Comment (konopka90) [inline apps/freeablo/engine/threadmanager.cpp]: Change 1000 to const

File: apps/freeablo/engine/threadmanager.cpp [hunk 5]
```diff
@@ -44,11 +44,11 @@ namespace Engine
...
             {
-                std::cout << "FPS: " << ((float)numFrames) / (((float)dur)/1000.0f) << std::endl;
+                std::cout << "FPS: " << ((float)numFrames) / (((float)duration)/1000.0f) << std::endl;
```
Comment (konopka90) [inline apps/freeablo/engine/threadmanager.cpp]: same here

**Comment by konopka90** [comment]
Changes:
   - engine/input.h moved to engine/engineinputmanager.h
   - ugly singleton EngineInputManager removed
   - Hotkey code improved
   - Bug fix in Launcher - couldn't find file with uppercase file extension.

**Comment by wheybags** [comment]
I'm happy to merge this if you think it's ready?

**Comment by konopka90** [comment]
Yeah sure

**Pull request merged**